### PR TITLE
Location path 2

### PR DESCRIPTION
**Pull request by petebacondarwin**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
fix

**What is the current behavior? (You can also link to an open issue here)**
see the commit message

**What is the new behavior (if this is a feature change)?**
see the commit message

**Does this PR introduce a breaking change?**
The previous behaviour would have led to a broken app, so it is highly
unlikely that any app relied on this behaviour. So this doesn't constitute
a breaking change.

**Please check if the PR fulfills these requirements**
- [ ] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Other information**:

Alternative solution #15359

### CODE REVIEW

File: src/ng/location.js [hunk 1]
```diff
@@ -144,9 +149,10 @@ function LocationHtml5Url(appBase, appBaseNoFile, basePrefix) {
     var appUrl, prevAppUrl;
     var rewrittenUrl;
 
+
     if (isDefined(appUrl = stripBaseUrl(appBase, url))) {
       prevAppUrl = appUrl;
-      if (isDefined(appUrl = stripBaseUrl(basePrefix, appUrl))) {
+      if (basePrefix && isDefined(appUrl = stripBaseUrl(basePrefix, appUrl))) {
```
Comment (petebacondarwin) [inline src/ng/location.js]: interestingly I think this was actually the cause of the bug in the first place...

Comment (mgol) [inline src/ng/location.js]: There are no tests for this change, are there?

File: test/ng/locationSpec.js [hunk 2]
```diff
@@ -2480,6 +2480,15 @@ describe('$location', function() {
       expect(parseLinkAndReturn(locationUrl, 'someIgnoredAbsoluteHref', '#test')).toEqual('http://server/pre/otherPath#test');
     });
 
+    it('should complain if the path starts with double slashes', function() {
+      expect(function() {
+        parseLinkAndReturn(locationUrl, 'http://server/pre///other/path');
+      }).toThrowMinErr('$location', 'badpath');
+
+      expect(function() {
+        parseLinkAndReturn(locationUrl, 'http://server/pre/\\\\other/path');
```
Comment (koto) [inline test/ng/locationSpec.js]: Could you add a test case with mixed slashes (`pre//\\//other/path`) ?

File: src/ng/location.js [hunk 3]
```diff
@@ -30,13 +30,18 @@ function parseAbsoluteUrl(absoluteUrl, locationObj) {
   locationObj.$$port = toInt(parsedUrl.port) || DEFAULT_PORTS[parsedUrl.protocol] || null;
 }
 
+var DOUBLE_SLASH_REGEX = /\s*[\\/]{2,}/;
+function parseAppUrl(url, locationObj) {
 
-function parseAppUrl(relativeUrl, locationObj) {
-  var prefixed = (relativeUrl.charAt(0) !== '/');
+  if (url.search(DOUBLE_SLASH_REGEX) === 0) {
```
Comment (gkalpak) [inline src/ng/location.js]: <sub>You could also add `^` at the beginning of `DOUBLE_SLASH_REGEX` and use  `DOUBLE_SLASH_REGEX.test(url)`?</sub>

**Review by gkalpak — APPROVED** [review]

File: test/ng/locationSpec.js (hunk 2 above)
Comment (petebacondarwin) [inline test/ng/locationSpec.js]: will do

File: src/ng/location.js (hunk 1 above)
Comment (petebacondarwin) [inline src/ng/location.js]: No new tests, specifically, but it doesn't break any of the current tests
And there are the new tests with the double slashes.
I couldn't think of a specific test for this change. Any ideas?

**Comment by petebacondarwin** [comment]
@koto and @mprobst - any further comments before we merge this?

**Comment by koto** [comment]
Looks good to me, security-wise.

**Comment by mprobst** [comment]
LGTM

**Comment by ivanrey** [comment]
This fix has problems with internal routing of an app. Sometimes double slash in fragments might be important to indicate an specific state.

I believe this fix should not take into account the fragment part of the URL since it's not sent to the server and breaks some applications.

**Comment by gkalpak** [comment]
@ivanrey, afaict this is supposed to only check for the beginning of the client-side URL and I think it didn't work correctly before either. Do you have a specific usecase/demo you can share, that used to work before and broke as a result of this fix?

**Pull request merged**